capsules/src/net/thread/tlv.rs

```diff
@@ -0,0 +1,1315 @@
+                offset = enc_consume!(buf, offset; encode_u8, parent_priority);
+                offset = enc_consume!(buf, offset; encode_u8, link_quality_3);
+                offset = enc_consume!(buf, offset; encode_u8, link_quality_2);
+                offset = enc_consume!(buf, offset; encode_u8, link_quality_1);
+                offset = enc_consume!(buf, offset; encode_u8, leader_cost);
+                offset = enc_consume!(buf, offset; encode_u8, id_sequence);
+                offset = enc_consume!(buf, offset; encode_u8, active_routers);
+                if let Some(ref buf_size) = sed_buffer_size {
+                    offset = enc_consume!(buf, offset; encode_u16, buf_size.to_be());
+                }
+                if let Some(ref datagram_cnt) = sed_datagram_count {
+                    offset = enc_consume!(buf, offset; encode_u8, *datagram_cnt);
+                }
+                stream_done!(offset)
+            }
+            Tlv::LinkMargin(ref link_margin) => {
+                let value_width = mem::size_of::<u8>();
+                let mut offset = enc_consume!(buf; self; encode_tl, value_width);
+                offset = enc_consume!(buf, offset; encode_u8, *link_margin);
+                stream_done!(offset)
+            }
+            Tlv::Status(ref status) => {
+                let value_width = mem::size_of::<u8>();
+                let mut offset = enc_consume!(buf; self; encode_tl, value_width);
+                offset = enc_consume!(buf, offset; encode_u8, *status);
+                stream_done!(offset)
+            }
+            Tlv::Version(ref version) => {
+                let value_width = mem::size_of::<u16>();
+                let mut offset = enc_consume!(buf; self; encode_tl, value_width);
+                offset = enc_consume!(buf, offset; encode_u16, *version);
+                stream_done!(offset)
+            }
+            Tlv::ActiveOperationalDataset(ref network_mgmt_tlvs) => {
+                let value_width = network_mgmt_tlvs.len();
+                let mut offset = enc_consume!(buf; self; encode_tl, value_width);
+                offset = enc_consume!(buf, offset; encode_bytes, network_mgmt_tlvs);
+                stream_done!(offset)
+            }
+            Tlv::PendingOperationalDataset(ref network_mgmt_tlvs) => {
+                let value_width = network_mgmt_tlvs.len();
+                let mut offset = enc_consume!(buf; self; encode_tl, value_width);
+                offset = enc_consume!(buf, offset; encode_bytes, network_mgmt_tlvs);
+                stream_done!(offset)
+            }
+        }
+    }
+
+    fn encode_tl(&self, buf: &mut [u8], value_width: usize) -> SResult {
+        stream_len_cond!(buf, TL_WIDTH + value_width);
+        buf[0] = TlvType::from(self) as u8;
+        buf[1] = value_width as u8;
+        stream_done!(TL_WIDTH)
+    }
+
+    pub fn decode(buf: &mut [u8]) -> SResult<Tlv> {
```

Comment:
"deserialize" if "encode -> decode". Are there any expectations for passed in buffer (e.g. "the buffer should contain a TLV header formatted following the specifications in blah blah blah")? Why is it `&mut` rather than `&`?